client/src/main/java/io/pravega/client/stream/impl/TransactionalEventStreamWriterImpl.java

```diff
@@ -0,0 +1,217 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.stream.impl;
+
+import com.google.common.base.Preconditions;
+import io.pravega.client.segment.impl.Segment;
+import io.pravega.client.segment.impl.SegmentOutputStream;
+import io.pravega.client.segment.impl.SegmentOutputStreamFactory;
+import io.pravega.client.stream.EventWriterConfig;
+import io.pravega.client.stream.Serializer;
+import io.pravega.client.stream.Stream;
+import io.pravega.client.stream.Transaction;
+import io.pravega.client.stream.Transaction.Status;
+import io.pravega.client.stream.TransactionalEventStreamWriter;
+import io.pravega.client.stream.TxnFailedException;
+import io.pravega.common.concurrent.ExecutorServiceHelpers;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+
+import static io.pravega.common.concurrent.Futures.getAndHandleExceptions;
+
+/**
+ * This class takes in events, finds out which segment they belong to and then calls write on the appropriate segment.
+ * It deals with segments that are sealed by re-sending the unacked events to the new correct segment.
+ * 
+ * @param <Type> The type of event that is sent
+ */
+@Slf4j
+@ToString(of = { "stream", "closed" })
+public class TransactionalEventStreamWriterImpl<Type> implements TransactionalEventStreamWriter<Type> {
+
+    private final Stream stream;
+    private final Serializer<Type> serializer;
+    private final SegmentOutputStreamFactory outputStreamFactory;
+    private final Controller controller;
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final EventWriterConfig config;
+    private final ExecutorService retransmitPool;
+    private final Pinger pinger;
+    
+    TransactionalEventStreamWriterImpl(Stream stream, Controller controller, SegmentOutputStreamFactory outputStreamFactory,
```

Comment:
I am avoiding that for the moment, because it has a bug in my IDE that breaks the syntax highlighting for the method. (Compiles OK though). I suspect I just need to update the plugin.